data_science/archive_exploration/01 - subject coocurrence.ipynb

```diff
@@ -0,0 +1,386 @@
+   "metadata": {},
+   "source": [
+    "At this point it's often helpful to index our data, ie transform words into numbers. We'll create two dictionaries which map back and forth between the subjects and their corresponding indicies:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "index_to_subject = {index: subject for index, subject in enumerate(subjects)}\n",
+    "subject_to_index = {subject: index for index, subject in enumerate(subjects)}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lets instantiate an empty numpy array which we'll then fill with our coocurrence data. Each column and each row will represent a subject - each cell (the intersection of a column and row) will therefore represent the 'strength' of the interaction between those subjects. As we haven't seen any interactions yet, we'll set every array element to 0."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "adjacency = np.empty((len(subjects), len(subjects)), \n",
+    "                     dtype=np.uint16)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To populate the matrix, we want to find every possible combination of subject in each sub-list from our original column, ie if we had the subjects\n",
+    "\n",
+    "`[Disease, Heart, Heart Diseases, Cardiology]`\n",
+    "\n",
+    "we would want to return \n",
+    "\n",
+    "`\n",
+    "[['Disease', 'Disease'],\n",
+    " ['Heart', 'Disease'],\n",
+    " ['Heart Diseases', 'Disease'],\n",
+    " ['Cardiology', 'Disease'],\n",
+    " ['Disease', 'Heart'],\n",
+    " ['Heart', 'Heart'],\n",
+    " ['Heart Diseases', 'Heart'],\n",
+    " ['Cardiology', 'Heart'],\n",
+    " ['Disease', 'Heart Diseases'],\n",
+    " ['Heart', 'Heart Diseases'],\n",
+    " ['Heart Diseases', 'Heart Diseases'],\n",
+    " ['Cardiology', 'Heart Diseases'],\n",
+    " ['Disease', 'Cardiology'],\n",
+    " ['Heart', 'Cardiology'],\n",
+    " ['Heart Diseases', 'Cardiology'],\n",
+    " ['Cardiology', 'Cardiology']]\n",
+    "`\n",
+    "\n",
+    "The `cartesian()` function which I've defined above will do that for us. We then find the appropriate intersection in the matrix and add another unit of 'strength' to it.  \n",
+    "We'll do this for every row of subjects in the `['Subjects']` column."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for row_of_subjects in tqdm(df['Subject'].dropna()):\n",
+    "    for subject_pair in cartesian(row_of_subjects, row_of_subjects):\n",
+    "        subject_index_1 = subject_to_index[clean(subject_pair[0])]\n",
+    "        subject_index_2 = subject_to_index[clean(subject_pair[1])]\n",
+    "\n",
+    "        adjacency[subject_index_1, subject_index_2] += 1"
```

Comment:
Either used 0-based indexing or 1-based indexing, but having `subject_index_1` and `subject_pair[1]` be unrelated is a recipe for confusion!

Also, this is something where itertools.product and tuple unpacking might come in handy. I think the following is equivalent, but cleaner:

```python
import itertools

for row_of_subjects in tqdm(df['Subject'].dropna()):
    for subject0, subject1 in itertools.product(row_of_subject, repeat=2):
        idx0 = subject_to_index[clean(subject0)]
        idx1 = subject_to_index[clean(subject1)]
        adjacency[idx0, idx1] += 1
```

Reply:
Good catch re: normalisation, and thanks for the `itertools.product()` pointer. I'm amazed that I didn't come across it when I first wrote that cartesian function!